### PR TITLE
feat: update MongoDB.Driver to 3.0.0

### PR DESCRIPTION
**Pull request by micdah**
In 3.0.0, all of Core has been merged into one, which is why we can drop dependency on `MongoDB.Driver.GridFS`.

`GuidRepresentation` has also been altered, so now to get a default Guid representation behaviour, we must register a global `GuidSerializer` with a selected representation.

Alternatively, a representation must be choosen on a per-field basis using annotations.

---
Rebus is [MIT-licensed](https://opensource.org/licenses/MIT). The code submitted in this pull request needs to carry the MIT license too. By leaving this text in, __I hereby acknowledge that the code submitted in the pull request has the MIT license and can be merged with the Rebus codebase__.

### CODE REVIEW

**Comment by micdah** [comment]
Well, there seems to be a potential breaking change here.

Specifically MongoDb.Driver 3.0 requires explicit specification of `GuidRepresentation`, either thorugh a `BsonClassMap` of annotations on the classes. 

So stuff like `BsonValue.Create(object)` when given a `Guid` will now throw, you must use stuff like `new BsonBinaryData(Guid, GuidRepresentation)`.

I tried playing around with just doing that - but it seems that `MongoDbSagaStorage` is sometimes given a `string` typed `propertyValue` for the `Id` field, where I would have assumed only a `Guid` would ever be a valid option.

So I'm unsure the best approach to handle this breaking change in Rebus.

We can't just set up a global `GuidSerializer` as that might interfere with / break customer code as the serializer is global.

We can't just add a `BsonClassMap` on `ISagaData` because the MongoDB driver requires the target type to be a class, not an interface. I tried applying it to `SagaData` just to see, but there were still a lot of places which brokke. Applying to the default implementation would also not work for any users implementing the `ISagaData` interface directly.

I guess the intended approach from MongoDB.Driver is that it must be made explicit, e.g. users of `Rebus.MongoDB` needs to ensure a `BsonClassMap` is registered for all concrete implementations of `ISagaData` in use - and then decide which `GuidRepresentation` to use (_and we'd need to know that choice somehow, as the `Find` etc., will need to serialise an appropriate `BsonBinaryData` to search for_) or that annotations are placed on all `Guid` properties (_which would not work if inheriting from `SagaData` which is missing those_).

**Comment by micdah** [comment]
Please note, even though this PR is green - it's not actually production ready.

This just demonstrates that we can upgrade to 3.0, but we need to figure out how to handle GuidRepresentation from now on.

I have just configured it such that we are assuming `CSharpLegacy` everywhere, which can be done i 3.0, but is not done so by default.

Clients may also mix and match as they see fit, e.g. we could have a `ISagaData` implementation where `_id` is using `CSharpLegacy` but it may also contain other Guid fields which could be using `Standard`, as such this code especially has issues with working in 3.x

```csharp
    public async Task<ISagaData> Find(Type sagaDataType, string propertyName, object propertyValue)
    {
        var collection = await GetCollection(sagaDataType);

        var value = propertyValue is Guid guid
            ? new BsonBinaryData(guid, GuidRepresentation.CSharpLegacy)
            : BsonValue.Create(propertyValue);

        var criteria = propertyName == nameof(ISagaData.Id)
            ? new BsonDocument { { "_id", value } }
            : new BsonDocument { { propertyName, value } };

        var result = await collection.Find(criteria).FirstOrDefaultAsync().ConfigureAwait(false);

        if (result == null) return null;

        return (ISagaData)BsonSerializer.Deserialize(result, sagaDataType);
    }
```

For `_id` we could enforce `CSharpLegacy` (_for backwards compatibility_), but the client may have more `Guid` fields on the concrete class, which may or may not be using that representation - so we'd need to know that on a per-field basis for the above to work in a general approach.

I think for 3.x, when using the LINQ approach or using `Mongo.Driver.Builder` it will infer the type from available annotations / class maps, but when we are using `BsonDocument` (untyped) we explicitly need to use the appropriate binary representation to match the stored document.

**Comment by AGiorgetti** [comment]
I think that enforcing `CSharpLegacy` is not a good idea, it will force new project behave in a "bad" way.

I think that the following scenarios should be supported:
- The users should map all their classes (this is how MongoDB driver is intended to be used anyway) and we should add a warning in the readme pointing to MongoDB changelog; there's no way to overcome this issue because it's a breking change of the driver. If you want to keep compatibility you have to properly map all the guid fields (or change the serializer globally).
- Rebus Mongodb have to be extended with a configuration settings that let the user decide the "GuidRepresentation" for all the manual bson document generation that happens in the code (like find method you reported above).
- defaults should be aligned to those of the mongodb driver and we should provide a breaking change section that explains how to configure things to keep compatibility.

**Comment by micdah** [comment]
> I think that enforcing `CSharpLegacy` is not a good idea, it will force new project behave in a "bad" way.
> 
> I think that the following scenarios should be supported:
> 
> * The users should map all their classes (this is how MongoDB driver is intended to be used anyway) and we should add a warning in the readme pointing to MongoDB changelog; there's no way to overcome this issue because it's a breking change of the driver. If you want to keep compatibility you have to properly map all the guid fields (or change the serializer globally).
> * Rebus Mongodb have to be extended with a configuration settings that let the user decide the "GuidRepresentation" for all the manual bson document generation that happens in the code (like find method you reported above).
> * defaults should be aligned to those of the mongodb driver and we should provide a breaking change section that explains how to configure things to keep compatibility.

I agree, and in my opinion the first option seems the most "correct" by leaning against what the developers of `MongoDB.Driver` intends - iff a user then opts to just apply a `GuidRepresentation` globally, that's fine and will also just work here.

The main pain point of that though, is the `MongoDbSagaStorage.Find` method I think, as it has to lookup a class map to find the appropriate serialization strategy to use for the given root property of the saga data type.

**Comment by micdah** [comment]
I have updated the PR with a proposed implementation.

This requires the user to setup appropriate `BsonClassMap`'s for all their implementations of saga data.

If they inherit from `SagaData`, they will need to register a map for this base class:
```csharp
BsonClassMap.RegisterClassMap<SagaData>(map =>
{
    map.MapIdMember(obj => obj.Id).SetSerializer(new GuidSerializer(GuidRepresentation.Standard));
    map.MapMember(obj => obj.Revision);
});
```

If their inheriting class also contains `Guid`'s, they will further also have to create a class map for it:
```csharp
BsonClassMap.RegisterClassMap<MySagaData>(map =>
{
    map.MapMember(obj => obj.MyGuidProperty).SetSerializer(new GuidSerializer(GuidRepresentation.Standard));
});
```

If they inhertit from `ISagaData`, they will have to create a class map for their own class including all members of `ISagaData` (_MongoDB driver only allows mapping of defined, not inherited, members - inherited members must be defined in a `BsonClassMap` for that specific class, which is applied to all classes inheriting for it_).

I opted to register a global `GuidSerializer` in the Rebus test suite, because I could not create class maps for the generic tests such as `ConcurrencyHandling<TestMongoDbSagaStorage>` as they contain private `ISagaData` implementations for which we cannot create a `BsonClassMap` here.

**Comment by micdah** [comment]
@mookid8000 thoughts on this approach? 

Would be awesome to get `Rebus.MongoDB` unblocked for `MongoDB.Driver 3.x` as we have a pending upgrade internally which is blocked by this package alone 🙈

**Comment by mookid8000** [comment]
Hey @micdah , thanks for pushing this 🙂 Would it be possible for you to give me a summary at this point?

I'd love to merge & release it quickly, but I am not using MongoDB together with Rebus anywhere at the moment, so I am out of touch with a bunch of thing and would not feel confident enough to go ahead and release the update without fully understanding the nature of the breaking change.

**Comment by micdah** [comment]
Hey @mookid8000 

Of course, let me try and outline what the change from `MongoDB.Driver` is all about.

So TL;DR; is that previously `Guid` where stored in what MongoDB calls `CSharpLegacy` representation, which is .Net specific, there are similar Java and Python specific representations.

In 3.x they are now changing so that you can (must) choose what representation you want and they recommend using `Standard` which is a newer cross-language representation.

But the crux of it, is that you must decide representation - so out of the box `Guid`'s cannot be persisted or hydrated unless you tell the driver what representation you want, on a per field level!

# Ways of choosing representation

## Globally

You can opt to just use the same representation across the board, by setting up a global serializer
```csharp
BsonSerializer.RegisterSerializer(new GuidSerializer(GuidRepresentation.Standard));
```

If one does that, it will work just as it did pre 3.x

For Rebus.MongoDB I don't think this is the way to go, as this is global and will affect all code using MongoDB so client application will be forced to use whichever representation we choose to setup here.


## Per field

The other way is setting up per-field, either as an attribute:
```csharp
[BsonGuidRepresentation(GuidRepresentation.Standard)]
public Guid Id { get; set; }
```

Or in a `BsonClassMap`:
```csharp
BsonClassMap.RegisterClassMap<SomeClass>(map =>
{
    map.MapIdMember(obj => obj.Id)
        .SetSerializer(new GuidSerializer(GuidRepresentation.CSharpLegacy));

});
```

# What I propose here

So in this PR I basically propose to let the users choose - but it's imperative that this is **breaking**, so `Rebus.MongoDB` will not work out of the box, without the user doing _something_.

One thing a user could do, if they **inherit from `SagaData`** is:
```csharp
        BsonClassMap.RegisterClassMap<SagaData>(map =>
        {
            map.MapIdMember(obj => obj.Id).SetSerializer(new GuidSerializer(GuidRepresentation.Standard));
            map.MapMember(obj => obj.Revision);
        });
```

If they have existing data using `CSharpLegacy` they will need to match this so that existing data can be fetched correctly as well.

If a user inherits from `ISagaData`, they will need to create maps for these classes as well - you cannot create a `BsonClassMap` for an interface, so only when inheriting from `SagaData` can they create a single map applied across the board.

If a user has other `Guid`'s in their own saga data classes, these will have to be mapped appropriately as well - but this is the same that a user must to when switching to 3.x anyway.

For us specifically we are probably going to go with a global `CSharpLegacy` approach, as all our existing data has this representation, and consistency is always king. 🙈 

I think there should be some kind of migration guide in the readme on this change guiding users in what they ought to do?

**Comment by micdah** [comment]
I just pushed a commit to `README.md` where I outline what a user of this package need to do in order to configure Guid representation.

It's quite a lot of words for something "so small", but it's also very important users understand the what and why's of it I think

**Comment by micdah** [comment]
Just FYI, we've made a pre-release nuget package version of this PR in our internal NuGet feed and is rolling out our production services against it - so far it's going fine.

Hopefully we can get this into the official version in the near future, so we can go back to the official nuget version 🙈

**Comment by pfab-io** [comment]
hello, any update? we can't upgrade to mongo 3 because of the dependencies of this package. just to have any idea of possible dates

**Comment by micdah** [comment]
> hello, any update? we can't upgrade to mongo 3 because of the dependencies of this package. just to have any idea of possible dates

Haven’t heard anything yet, but as mentioned we are running with a pre-release version built on this PR and everything seems to be fine.

**Comment by mookid8000** [comment]
Hello everyone, sorry you had you wait 😅 I've now finally found time to understand what this was about, and thanks to @micdah 's thorough work I could simply accept the PR and build & push the NuGet.

It's our as Rebus.MongoDb 9.0.0 on NuGet.org now!

Thanks for contributing ❤️
